src/app/customizable-demo-form.component.ts

```diff
@@ -195,13 +271,34 @@ export class CustomizableDemoForm {
     this.sliderScoreThresholds[2] = (1 + this.sliderScoreThresholds[1]) / 2;
   }
 
+  onDemoSettingsChanged() {
+    let newDemoSettings = this.getDemoSettings();
+    if (JSON.stringify(this.demoSettings) !== JSON.stringify(newDemoSettings)) {
+      console.debug('Updating this.demoSettings', newDemoSettings);
+      this.demoSettings = newDemoSettings;
+
+      const encodedUISettings =
+        encodeURIComponent(JSON.stringify(this.getUISettings()));
+      const encodedDemoSettings =
+        encodeURIComponent(JSON.stringify(this.demoSettings));
+
+      this.router.navigate(
+        ['/customize', encodedUISettings, encodedDemoSettings]);
+    } else {
+      console.debug(
+        'Calling onDemoSettingsChanged(), but settings are unchanged',
+        newDemoSettings);
+    }
+  }
+
   /**
-   * Use a function to get the demo settings rather than storing it in a member
-   * variable because Angular will only call change detection if the entire
-   * object has changed, and will not listen to sub-properties by default.
+   * Note: This function must be called to update this.demoSettings for Angular
+   * to notice the change, because Angular will only call change detection if
+   * the entire object has changed, and will not listen to sub-properties by
+   * default.
    */
-  getDemoSettings() {
-    return {
+  private getDemoSettings(): DemoSettings {
+    return JSON.parse(JSON.stringify({
```

Comment:
What does `JSON.parse(JSON.stringify(` accomplish?

Reply:
Deep copy of the object (no accidental sharing of references to internal arrays and such)